libslide/src/parser.rs

```diff
@@ -0,0 +1,210 @@
+        self.add_sub_term()
+    }
+
+    fn add_sub_term(&mut self) -> Box<Expr> {
+        let lhs = self.mul_divide_mod_term();
+        match self.token().token_type {
+            TokenType::Plus | TokenType::Minus => {
+                let operand = self.token().clone();
+                self.advance();
+                Box::new(Expr::BinOp(BinOp {
+                    op: operand,
+                    lhs,
+                    rhs: self.mul_divide_mod_term(),
+                }))
+            }
+            TokenType::CloseParen => lhs,
+            TokenType::CloseBracket => lhs,
+            TokenType::EOF => lhs,
+            _ => unreachable!(),
+        }
+    }
+
+    fn mul_divide_mod_term(&mut self) -> Box<Expr> {
+        let lhs = self.exponent_term();
+        match self.token().token_type {
+            TokenType::Mult | TokenType::Div | TokenType::Mod => {
+                let operand = self.token().clone();
+                self.advance();
+                Box::new(Expr::BinOp(BinOp {
+                    op: operand,
+                    lhs,
+                    rhs: self.exponent_term(),
+                }))
+            }
+            _ => lhs,
+        }
+    }
+
+    fn exponent_term(&mut self) -> Box<Expr> {
+        let lhs = self.num_term();
+        match self.token().token_type {
+            TokenType::Exp => {
+                let operand = self.token().clone();
+                self.advance();
+                Box::new(Expr::BinOp(BinOp {
+                    op: operand,
+                    lhs,
+                    rhs: self.exponent_term(),
+                }))
+            }
+            _ => lhs,
+        }
+    }
+
+    fn num_term(&mut self) -> Box<Expr> {
+        let node = match self.token().token_type {
+            TokenType::Plus | TokenType::Minus => {
+                let operand = self.token().clone();
+                self.advance();
+                let node = Box::new(Expr::UnaryOp(UnaryOp {
+                    op: operand,
+                    rhs: self.exponent_term(),
+                }));
+                node
+            }
+            TokenType::Float(f) => {
+                let node = Box::new(Expr::Float(f));
+                self.advance();
+                node
+            }
+            TokenType::Int(i) => {
+                let node = Box::new(Expr::Int(i));
+                self.advance();
+                node
+            }
+            TokenType::OpenParen => {
+                // eat left paren
+                self.advance();
+                let node = self.expr();
+                self.advance();
+                node
+            }
+            TokenType::OpenBracket => {
+                // eat left bracket
+                self.advance();
+                let node = self.expr();
+                self.advance();
+                node
+            }
+            TokenType::Variable(_) => {
+                self.assignment()
+            }
```

Comment:
But this won't work if we do something like `1 + a` now?

Reply:
LGTM other than that, great job :+1: but we should fix this

Reply:
Don't worry about robustness for now, we can just panic wherever, but we need to be able to support this so that we can start the optimization pipeline

Reply:
If you want I can do it, just lmk

Reply:
Other option is we can ignore this and just say we only parse one expression/statement for now, since that's what we do anyway. Otherwise we'll need to support multi-statement programs for variables in the RHS to be useful anyway. That is, with this parser

```
a = 1;
b = 1 + a;
```

is not supported anyway because we parse at most one expression.

Reply:
Let me know what you think. I'm happy to just leave as is for now so we can start on the other parts of this program, and we can clean this up later.

Reply:
@luke-bhan

Reply:
sounds good. Lets just leave it. Ready to merge I think right?

Reply:
I also completely forgot about that lol

Reply:
@ayazhafiz